File: backend/src/main/java/com/bryan/system/service/user/UserService.java
package com.bryan.system.service.user;

import com.bryan.system.domain.dto.UserUpdateDTO;
import com.bryan.system.domain.entity.SysUser;
import com.bryan.system.domain.entity.UserRole;
import com.bryan.system.domain.enums.user.UserStatusEnum;
import com.bryan.system.domain.request.user.ChangeRoleRequest;
import com.bryan.system.domain.request.user.UserCreateRequest;
import com.bryan.system.domain.request.user.UserSearchRequest;
import com.bryan.system.domain.response.PageResult;
import com.bryan.system.exception.BusinessException;
import com.bryan.system.exception.ResourceNotFoundException;
import com.bryan.system.mapper.UserMapper;
import com.bryan.system.util.jwt.JwtUtils;
import lombok.RequiredArgsConstructor;
import lombok.extern.slf4j.Slf4j;
import org.springframework.security.crypto.password.PasswordEncoder;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.time.LocalDateTime;
import java.util.ArrayList;
import java.util.List;
import java.util.Optional;
import java.util.Set;
import java.util.stream.Collectors;

/**
 * 用户服务实现类，处理用户注册、登录、信息管理、导出等业务逻辑。
 *
 * @author Bryan Long
 */
@Slf4j
@Service
@RequiredArgsConstructor
public class UserService {

    private final UserMapper userMapper;
    private final PasswordEncoder passwordEncoder;
    private final UserRoleService userRoleService;

    /**
     * 管理员创建用户
     *
     * @param req 创建请求
     * @return 新增的用户实体
     */
    @Transactional
    public SysUser createUser(UserCreateRequest req) {
        if (userMapper.selectByUsername(req.getUsername()) != null) {
            throw new BusinessException("用户名已存在");
        }

        List<Long> roleIds = req.getRoleIds() == null
                ? new ArrayList<>()
                : new ArrayList<>(req.getRoleIds());
        if (roleIds.isEmpty()) {
            UserRole defaultRole = userRoleService.getDefaultRole();
            if (defaultRole == null) {
                throw new BusinessException("系统未配置默认角色");
            }
            roleIds.add(defaultRole.getId());
        }

        List<UserRole> roles = userRoleService.listByIds(roleIds);
        if (roles.size() != roleIds.size()) {
            Set<Long> exist = roles.stream()
                    .map(UserRole::getId)
                    .collect(Collectors.toSet());
            List<Long> missing = new ArrayList<>(roleIds);
            missing.removeAll(exist);
            throw new IllegalArgumentException("角色不存在：" + missing);
        }

        String roleNames = roles.stream()
                .map(UserRole::getRoleName)
                .collect(Collectors.joining(","));

        LocalDateTime now = LocalDateTime.now();

        SysUser sysUser = SysUser.builder()
                .username(req.getUsername())
                .password(passwordEncoder.encode(req.getPassword()))
                .phone(req.getPhone())
                .email(req.getEmail())
                .roles(roleNames)
                .status(UserStatusEnum.NORMAL)
                .loginFailCount(0)
                .passwordResetAt(now)
                .build();
        this.fillInsert(sysUser);

        int saved = userMapper.insert(sysUser);
        if (saved == 0) {
            throw new BusinessException("插入数据库失败");
        }

        log.info("管理员创建用户成功: id: {}, username: {} ", sysUser.getId(), sysUser.getUsername());
        return sysUser;
    }

    /**
     * 获取所有用户列表（不分页）。
     *
     * @return 包含所有用户的分页对象（Page）。
     */
    public PageResult<SysUser> getAllUsers(int pageNum,
                                           int pageSize) {
        int offset = (pageNum - 1) * pageSize;
        List<SysUser> rows = userMapper.selectPage(
                offset,
                pageSize,
                null);
        long total = userMapper.count(null);
        return PageResult.of(rows, pageNum, pageSize, total);
    }

    /**
     * 根据用户ID获取用户信息。
     *
     * @param userId 用户 ID
     * @return 用户实体对象
     * @throws ResourceNotFoundException 用户不存在时抛出
     */
    public SysUser getUserById(Long userId) {
        return Optional.ofNullable(userMapper.selectById(userId))
                .orElseThrow(() -> new ResourceNotFoundException("用户不存在"));
    }

    /**
     * 根据用户名获取用户信息。
     *
     * @param username 用户名
     * @return 用户实体对象
     */
    public SysUser getUserByUsername(String username) {
        return userMapper.selectByUsername(username);
    }

    /**
     * 通用用户搜索，支持多条件模糊查询和分页。
     *
     * @param searchRequest 搜索请求
     * @return 符合查询条件的分页对象（Page）
     */
    public PageResult<SysUser> queryUsers(UserSearchRequest searchRequest,
                                           int pageNum,
                                           int pageSize) {
        int offset = (pageNum - 1) * pageSize;
        List<SysUser> rows = userMapper.selectPage(
                offset,
                pageSize,
                searchRequest);
        long total = userMapper.count(searchRequest);
        return PageResult.of(rows, pageNum, pageSize, total);
    }

    /**
     * 根据用户 ID 列表批量查询用户信息
     *
     * @param userIds 用户 ID 列表
     * @return 用户信息列表
     */
    public List<SysUser> getUsersByIds(List<Long> userIds) {
        if (userIds == null || userIds.isEmpty()) {
            return new ArrayList<>();
        }
        return userMapper.selectByIdList(userIds);
    }

    /**
     * 检查用户是否存在
     *
     * @param userId 用户 ID
     * @return 用户是否存在
     */
    public boolean existsById(Long userId) {
        return userMapper.selectById(userId) != null;
    }

    /**
     * 更新用户基础信息（用户名和邮箱）。
     *
     * @param userId                     用户 ID
     * @param dto                        用户更新 DTO
     * @return                           更新后的用户对象
     * @throws ResourceNotFoundException 用户不存在时抛出
     * @throws BusinessException         用户名重复时抛出
     */
    public SysUser updateUser(Long userId, UserUpdateDTO dto) {
        SysUser user = this.getUserById(userId);
        if (user == null) {
            throw new BusinessException("用户不存在: " + userId);
        }

        if (dto.getPhone() != null) {
            user.setPhone(dto.getPhone());
        }
        if (dto.getEmail() != null) {
            user.setEmail(dto.getEmail());
        }

        this.fillUpdate(user);

        int rows = userMapper.update(user);
        if (rows != 1) {
            throw new BusinessException("更新失败，可能数据已变更");
        }

        log.info("用户ID: {} 的信息更新成功", userId);
        return user;
    }

    /**
     * 修改用户角色。
     *
     * @param userId 用户 ID
     * @param req  新角色字符串（多个角色用逗号分隔）
     * @return 更新后的用户对象
     * @throws ResourceNotFoundException 用户不存在时抛出
     */
    @Transactional
    public SysUser changeRoleByIds(Long userId, ChangeRoleRequest req) {
        List<Long> ids = req.getRoleIds();
        List<UserRole> roles = userRoleService.listByIds(ids);

        if (roles.size() != ids.size()) {
            Set<Long> exist = roles.stream()
                    .map(UserRole::getId)
                    .collect(Collectors.toSet());
            ids.removeAll(exist);
            throw new IllegalArgumentException("角色不存在：" + ids);
        }
        String roleNames = roles.stream()
                .map(UserRole::getRoleName)
                .collect(Collectors.joining(","));
        SysUser user = this.getUserById(userId);
        user.setRoles(roleNames);

        this.fillUpdate(user);

        userMapper.update(user);
        return user;
    }

    /**
     * 重置用户密码（管理员）。
     *
     * @param userId      用户 ID
     * @param newPassword 新密码（明文）
     * @return 更新后的用户对象
     * @throws ResourceNotFoundException 用户不存在时抛出
     * @throws BusinessException         旧密码验证失败时抛出
     */
    public SysUser resetPassword(Long userId, String newPassword) {
        SysUser user = this.getUserById(userId);
        user.setPassword(passwordEncoder.encode(newPassword));
        user.setPasswordResetAt(LocalDateTime.now());

        this.fillUpdate(user);

        userMapper.update(user);
        log.info("用户ID: {} 的密码强制修改成功", userId);
        return user;
    }

    /**
     * 封禁指定用户。
     *
     * @param userId 用户 ID
     * @return 更新后的用户对象
     * @throws ResourceNotFoundException 用户不存在时抛出
     */
    public SysUser blockUser(Long userId) {
        SysUser user = this.getUserById(userId);
        user.setStatus(UserStatusEnum.BANNED);

        this.fillUpdate(user);

        userMapper.update(user);
        log.info("用户ID: {} 封禁成功", userId);
        return user;
    }

    /**
     * 解封指定用户。
     *
     * @param userId 用户 ID
     * @return 更新后的用户对象
     * @throws ResourceNotFoundException 用户不存在时抛出
     */
    public SysUser unblockUser(Long userId) {
        SysUser user = this.getUserById(userId);
        user.setStatus(UserStatusEnum.NORMAL);

        this.fillUpdate(user);

        userMapper.update(user);
        log.info("用户ID: {} 解封成功", userId);
        return user;
    }

    /**
     * 删除用户（逻辑删除）。
     *
     * @param userId 用户 ID
     * @return 被删除的用户对象
     * @throws ResourceNotFoundException 用户不存在时抛出
     */
    public Long deleteUser(Long userId) {
        int rows = userMapper.deleteById(userId, LocalDateTime.now(), JwtUtils.getCurrentUsername());
        if (rows == 0) {
            throw new ResourceNotFoundException("用户不存在或已被删除");
        }
        log.info("用户ID: {} 删除成功 (逻辑删除)", userId);
        return userId;
    }

    private void fillInsert(SysUser user) {
        LocalDateTime now = LocalDateTime.now();
        Long operator = JwtUtils.getCurrentUserId();

        user.setDeleted(0);
        user.setVersion(0);
        user.setCreatedAt(now);
        user.setUpdatedAt(now);
        user.setUpdatedBy(operator.toString());
        user.setCreatedBy(operator.toString());

    }

    private  void fillUpdate(SysUser user) {
        LocalDateTime now = LocalDateTime.now();
        Long operator = JwtUtils.getCurrentUserId();

        user.setVersion(user.getVersion() + 1);
        user.setUpdatedAt(now);
        user.setUpdatedBy(operator.toString());
    }
}

File: go/user-system/internal/service/user_service.go
package service

import (
	"context"
	"fmt"
	"strings"

	"github.com/bryan/user-system/internal/model"
	"github.com/bryan/user-system/internal/pkg/jwt"
	"github.com/bryan/user-system/internal/pkg/response"
	"github.com/bryan/user-system/internal/repository"
	"go.uber.org/zap"
	"golang.org/x/crypto/bcrypt"
)

type UserService struct {
	userRepo    *repository.UserRepository
	roleSvc     *UserRoleService
	logger      *zap.Logger
}

func NewUserService(
	userRepo *repository.UserRepository,
	roleSvc *UserRoleService,
	logger *zap.Logger,
) *UserService {
	return &UserService{
		userRepo: userRepo,
		roleSvc:  roleSvc,
		logger:   logger,
	}
}

// CreateUser 管理员创建用户
func (s *UserService) CreateUser(ctx context.Context, req *model.UserCreateRequest) (*model.SysUser, error) {
	existing, _ := s.userRepo.SelectByUsername(req.Username)
	if existing != nil {
		return nil, response.NewBusinessException("用户名已存在")
	}

	roleIDs := req.RoleIDs
	if len(roleIDs) == 0 {
		defaultRole, err := s.roleSvc.GetDefaultRole(ctx)
		if err != nil || defaultRole == nil {
			return nil, response.NewBusinessException("系统未配置默认角色")
		}
		roleIDs = []int{defaultRole.ID}
	}

	roles, err := s.roleSvc.ListByIDs(ctx, roleIDs)
	if err != nil {
		return nil, response.NewBusinessException("角色查询失败")
	}
	if len(roles) != len(roleIDs) {
		return nil, response.NewBusinessException("部分角色不存在")
	}

	roleNames := make([]string, 0, len(roles))
	for _, r := range roles {
		roleNames = append(roleNames, r.RoleName)
	}
	rolesStr := strings.Join(roleNames, ",")

	hashedPwd, err := bcrypt.GenerateFromPassword([]byte(req.Password), bcrypt.DefaultCost)
	if err != nil {
		return nil, response.NewBusinessException("密码加密失败")
	}

	operator := getCurrentOperator(ctx)
	user := &model.SysUser{
		Username:        req.Username,
		Password:        string(hashedPwd),
		Phone:          req.Phone,
		Email:          req.Email,
		Roles:          &rolesStr,
		Status:         model.UserStatusNormal,
		Deleted:        0,
		Version:        0,
		LoginFailCount: intPtr(0),
		CreatedBy:      &operator,
		UpdatedBy:      &operator,
	}

	if err := s.userRepo.Insert(user); err != nil {
		return nil, response.NewBusinessException("插入数据库失败")
	}

	s.logger.Info("管理员创建用户成功", zap.Int64("id", user.ID), zap.String("username", user.Username))
	return user, nil
}

// GetAllUsers 获取所有用户列表（分页）
func (s *UserService) GetAllUsers(ctx context.Context, pageNum, pageSize int) (*response.PageResult, error) {
	offset := (pageNum - 1) * pageSize
	users, err := s.userRepo.SelectPage(offset, pageSize, nil)
	if err != nil {
		return nil, response.NewBusinessException("查询用户列表失败")
	}
	total, _ := s.userRepo.Count(nil)

	rows := make([]interface{}, 0, len(users))
	for i := range users {
		rows = append(rows, users[i])
	}
	pr := response.NewPageResult(rows, total, pageNum, pageSize)
	return &pr, nil
}

// GetUserByID 根据用户 ID 获取用户信息
func (s *UserService) GetUserByID(ctx context.Context, userID int64) (*model.SysUser, error) {
	user, err := s.userRepo.SelectByID(userID)
	if err != nil {
		return nil, response.NewResourceNotFoundException("用户不存在")
	}
	return user, nil
}

// GetUserByUsername 根据用户名获取用户信息
func (s *UserService) GetUserByUsername(ctx context.Context, username string) (*model.SysUser, error) {
	user, err := s.userRepo.SelectByUsername(username)
	if err != nil {
		return nil, err
	}
	return user, nil
}

// QueryUsers 通用用户搜索
func (s *UserService) QueryUsers(ctx context.Context, req *model.UserSearchRequest, pageNum, pageSize int) (*response.PageResult, error) {
	offset := (pageNum - 1) * pageSize
	users, err := s.userRepo.SelectPage(offset, pageSize, req)
	if err != nil {
		return nil, response.NewBusinessException("查询用户失败")
	}
	total, _ := s.userRepo.Count(req)

	rows := make([]interface{}, 0, len(users))
	for i := range users {
		rows = append(rows, users[i])
	}
	pr := response.NewPageResult(rows, total, pageNum, pageSize)
	return &pr, nil
}

// UpdateUser 更新用户基础信息
func (s *UserService) UpdateUser(ctx context.Context, userID int64, req *model.UserUpdateRequest) (*model.SysUser, error) {
	user, err := s.GetUserByID(ctx, userID)
	if err != nil {
		return nil, err
	}

	if req.Phone != nil {
		user.Phone = req.Phone
	}
	if req.Email != nil {
		user.Email = req.Email
	}

	user.Version++
	user.UpdatedBy = strPtr(getCurrentOperator(ctx))

	if err := s.userRepo.Update(user); err != nil {
		return nil, response.NewBusinessException("更新失败，可能数据已变更")
	}

	s.logger.Info("用户信息更新成功", zap.Int64("userId", userID))
	return user, nil
}

// ChangeRoleByIds 修改用户角色
func (s *UserService) ChangeRoleByIds(ctx context.Context, userID int64, req *model.ChangeRoleRequest) (*model.SysUser, error) {
	roles, err := s.roleSvc.ListByIDs(ctx, req.RoleIDs)
	if err != nil {
		return nil, response.NewBusinessException("角色查询失败")
	}
	if len(roles) != len(req.RoleIDs) {
		return nil, response.NewBusinessException("部分角色不存在")
	}

	roleNames := make([]string, 0, len(roles))
	for _, r := range roles {
		roleNames = append(roleNames, r.RoleName)
	}
	rolesStr := strings.Join(roleNames, ",")

	user, err := s.GetUserByID(ctx, userID)
	if err != nil {
		return nil, err
	}
	user.Roles = &rolesStr
	user.Version++
	user.UpdatedBy = strPtr(getCurrentOperator(ctx))

	if err := s.userRepo.Update(user); err != nil {
		return nil, response.NewBusinessException("角色更新失败")
	}
	return user, nil
}

// ResetPassword 重置用户密码（管理员）
func (s *UserService) ResetPassword(ctx context.Context, userID int64, newPassword string) (*model.SysUser, error) {
	user, err := s.GetUserByID(ctx, userID)
	if err != nil {
		return nil, err
	}

	hashedPwd, err := bcrypt.GenerateFromPassword([]byte(newPassword), bcrypt.DefaultCost)
	if err != nil {
		return nil, response.NewBusinessException("密码加密失败")
	}

	user.Password = string(hashedPwd)
	user.Version++
	user.UpdatedBy = strPtr(getCurrentOperator(ctx))

	if err := s.userRepo.Update(user); err != nil {
		return nil, response.NewBusinessException("密码重置失败")
	}

	s.logger.Info("用户密码强制修改成功", zap.Int64("userId", userID))
	return user, nil
}

// BlockUser 封禁指定用户
func (s *UserService) BlockUser(ctx context.Context, userID int64) (*model.SysUser, error) {
	user, err := s.GetUserByID(ctx, userID)
	if err != nil {
		return nil, err
	}
	user.Status = model.UserStatusBanned
	user.Version++
	user.UpdatedBy = strPtr(getCurrentOperator(ctx))

	if err := s.userRepo.Update(user); err != nil {
		return nil, response.NewBusinessException("封禁失败")
	}

	s.logger.Info("用户封禁成功", zap.Int64("userId", userID))
	return user, nil
}

// UnblockUser 解封指定用户
func (s *UserService) UnblockUser(ctx context.Context, userID int64) (*model.SysUser, error) {
	user, err := s.GetUserByID(ctx, userID)
	if err != nil {
		return nil, err
	}
	user.Status = model.UserStatusNormal
	user.Version++
	user.UpdatedBy = strPtr(getCurrentOperator(ctx))

	if err := s.userRepo.Update(user); err != nil {
		return nil, response.NewBusinessException("解封失败")
	}

	s.logger.Info("用户解封成功", zap.Int64("userId", userID))
	return user, nil
}

// DeleteUser 删除用户（逻辑删除）
func (s *UserService) DeleteUser(ctx context.Context, userID int64) (int64, error) {
	operator := getCurrentOperator(ctx)
	if err := s.userRepo.DeleteByID(userID, operator); err != nil {
		return 0, response.NewResourceNotFoundException("用户不存在或已被删除")
	}
	s.logger.Info("用户删除成功 (逻辑删除)", zap.Int64("userId", userID))
	return userID, nil
}

// getCurrentOperator 获取当前操作者标识
func getCurrentOperator(ctx context.Context) string {
	claims, ok := ctx.Value("claims").(*jwt.Claims)
	if !ok || claims == nil {
		return "SYSTEM"
	}
	return fmt.Sprintf("%d", claims.UserID)
}
